sdk/servicebus/Microsoft.Azure.ServiceBus/src/Core/MessageReceiver.cs

```diff
@@ -899,6 +903,41 @@ public void RegisterMessageHandler(Func<Message, CancellationToken, Task> handle
             this.OnMessageHandler(messageHandlerOptions, handler);
         }
 
+        /// <summary>
+        /// Unregister message handler from the receiver if there is an active message handler registered. This operation waits for the completion
+        /// of inflight receive and message handling operations to finish and unregisters future receives on the message handler which previously 
+        /// registered. 
+        /// </summary>
+        public async Task UnregisterMessageHandlerAsync()
+        {
+            this.ThrowIfClosed();
+
+            MessagingEventSource.Log.UnregisterMessageHandlerStart(this.ClientId);
+            lock (this.messageReceivePumpSyncLock)
+            {
+                if (this.receivePump == null || this.receivePumpCancellationTokenSource.IsCancellationRequested)
+                {
+                    // Silently return if handler has already been unregistered. 
+                    return;
+                }
+
+                this.receivePumpCancellationTokenSource.Cancel();
+                this.receivePumpCancellationTokenSource.Dispose();
+            }
+
+            while (this.receivePump != null
```

Comment:
Should hold the messageReceivePumpSyncLock while executing this loop to prevent new message handler registrations to happen as that would put the receiver in a bad state.